splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/BroadcastJoinStrategy.java

```diff
@@ -157,16 +132,28 @@ public void estimateCost(Optimizable innerTable,
             return; //actually a scan, don't do anything
         }
         innerCost.setBase(innerCost.cloneMe());
-        double joinSelectivity = SelectivityUtil.estimateJoinSelectivity(innerTable, cd, predList, (long) innerCost.rowCount(), (long) outerCost.rowCount(), outerCost);
-        double totalOutputRows = SelectivityUtil.getTotalRows(joinSelectivity, outerCost.rowCount(), innerCost.rowCount());
-        innerCost.setNumPartitions(outerCost.partitionCount());
-        double joinCost = SelectivityUtil.broadcastJoinStrategyLocalCost(innerCost, outerCost);
-        innerCost.setLocalCost(joinCost);
-        innerCost.setLocalCostPerPartition(joinCost);
-        innerCost.setRemoteCost(SelectivityUtil.getTotalRemoteCost(innerCost,outerCost,totalOutputRows));
-        innerCost.setRowOrdering(outerCost.getRowOrdering());
-        innerCost.setRowCount(totalOutputRows);
-        innerCost.setEstimatedHeapSize((long)SelectivityUtil.getTotalHeapSize(innerCost,outerCost,totalOutputRows));
+        double estimatedMemoryMB = innerCost.getEstimatedHeapSize()/1024d/1024d;
+        double estimatedRowCount = innerCost.getEstimatedRowCount();
+        SConfiguration configuration=EngineDriver.driver().getConfiguration();
+        long regionThreshold = configuration.getBroadcastRegionMbThreshold();
+        long rowCountThreshold = configuration.getBroadcastRegionRowThreshold();
+        if (estimatedMemoryMB<regionThreshold && estimatedRowCount<rowCountThreshold) {
+            double joinSelectivity = SelectivityUtil.estimateJoinSelectivity(innerTable, cd, predList, (long) innerCost.rowCount(), (long) outerCost.rowCount(), outerCost);
+            double totalOutputRows = SelectivityUtil.getTotalRows(joinSelectivity, outerCost.rowCount(), innerCost.rowCount());
+            innerCost.setNumPartitions(outerCost.partitionCount());
+            double joinCost = SelectivityUtil.broadcastJoinStrategyLocalCost(innerCost, outerCost);
+            innerCost.setLocalCost(joinCost);
+            innerCost.setLocalCostPerPartition(joinCost);
+            innerCost.setRemoteCost(SelectivityUtil.getTotalRemoteCost(innerCost, outerCost, totalOutputRows));
+            innerCost.setRowOrdering(outerCost.getRowOrdering());
+            innerCost.setRowCount(totalOutputRows);
+            innerCost.setEstimatedHeapSize((long) SelectivityUtil.getTotalHeapSize(innerCost, outerCost, totalOutputRows));
+        }
+        else {
+            // Set cost to max to rule out broadcast join
+            innerCost.setLocalCost(Double.MAX_VALUE);
+            innerCost.setRemoteCost(Double.MAX_VALUE);
```

Comment:
If it is hinted to broadcast, make it feasible and return actual cost. Otherwise, bump up the cost to rule it out. 

@dgomezferro take another look